osbuild/util/linux.py

```diff
@@ -274,3 +281,237 @@ def bits(n):
     }
 
     return res
+
+
+def fcntl_flock(fd: int, lock_type: int, wait: bool = False):
+    """Perform File-locking Operation
+
+    This function performs a linux file-locking operation on the specified
+    file-descriptor. The specific type of lock must be specified by the caller.
+    This function does not allow to specify the byte-range of the file to lock.
+    Instead, it always applies the lock operations to the entire file.
+
+    For system-level documentation, see the `fcntl(2)` man-page, especially the
+    section about `struct flock` and the locking commands.
+
+    This function always uses the open-file-description locks provided by
+    modern linux kernels. This means, locks are tied to the
+    open-file-description. That is, they are shared between duplicated
+    file-descriptors. Furthermore, acquiring a lock while already holding a
+    lock will update the lock to the new specified lock type, rather than
+    acquiring a new lock.
+
+    If `wait` is `False` a non-blocking operation is performed. In case the lock
+    is contested a `BlockingIOError` is raised by the python standard library.
+    If `Wait` is `True`, the kernel will suspend execution until the lock is
+    acquired.
+
+    If a synchronous exception is raised, the operation will be canceled and the
+    exception is forwarded.
+
+    Parameters
+    ----------
+    fd
+        The file-descriptor to use for the locking operation.
+    lock_type
+        The type of lock to use. This can be one of: `fcntl.F_RDLCK`,
+        `fcntl.F_WRLCK`, `fcntl.F_UNLCK`.
+    wait
+        Whether to suspend execution until the lock is acquired in case of
+        contested locks.
+
+    Raises
+    ------
+    OSError
+        If the underlying `fcntl(2)` syscall fails, a matching `OSError` is
+        raised. In particular, `BlockingIOError` signals contested locks. The
+        POSIX error code is `EAGAIN`.
+    """
+
+    valid_types = [fcntl.F_RDLCK, fcntl.F_WRLCK, fcntl.F_UNLCK]
+    if not isinstance(fd, int) or fd < 0 or not any(lock_type == v for v in valid_types):
+        raise ValueError()
```

Comment:
I'd like a message on which of the cases went wrong, passing a wrong type, an errored fd is different from requesting an invalid lock type.